feat(eslint-plugin-react-hooks): support flat config

michaelfaith opened the pull request:
I also updated the README to include usage examples.

Example usage:
```js
import reactHooks from 'eslint-plugin-react-hooks';

export default [
 {
    files: ['**/*.{js,jsx}'],
    languageOptions: {
      ecmaVersion: 2020,
      globals: globals.browser,
      parserOptions: {
        ecmaVersion: 'latest',
        ecmaFeatures: {jsx: true},
        sourceType: 'module',
      },
    },
    settings: {react: {version: '18.3'}},
    ...reactHooks.configs['recommended-latest'],
  },
];
```

Closes #28313

Byron2016 commented:
Hi, any advance with this?

jmadson-chwy commented:
Looks like the most recent commit failed for this PR due to a lint issue. How ironic lol

michaelfaith commented:
> Hi, any advance with this?

Not sure.  Waiting on someone from the team to look at it.

eps1lon commented on packages/eslint-plugin-react-hooks/src/index.js:
```diff
@@ -9,18 +9,45 @@
 
 import RulesOfHooks from './RulesOfHooks';
 import ExhaustiveDeps from './ExhaustiveDeps';
+import {name, version} from '../package.json';
```
Can we `require` this instead to avoid having to use the rollup plugin? 

Alternatively, you could also hardcode the name and use `version` from `~/Reactversions.js`. But let's try `require` first.

michaelfaith replied:
Updated

eps1lon commented:
Can you rebase? Seems like CI missed this.

michaelfaith commented:
> Can you rebase? Seems like CI missed this.

Done!

eps1lon commented:
The support for flat config is backwards compatible with ESLint v8 and lower?

michaelfaith commented:
> The support for flat config is backwards compatible with ESLint v8 and lower?

Yep!  This shouldn't impact any earlier versions.  The "legacy" configs are exported from the same place as before (`configs`), while the new flat configs are on a new object `flatConfigs`.  Older configs (rc-based), will just be looking for the presence of the configs, and won't care or notice that a new export is present, while flat config users will just interact with the new export.  The new `meta` export is needed for flat config support, but won't interfere with older versions.

JstnMcBrd commented:
> the new flat configs are on a new object `flatConfigs`

🚨 WARNING 🚨

Putting flat configs on a new `flatConfig` field will make ESLint type definitions throw errors because that does not follow the correct ESLint plugin shape. This will disrupt some users with TypeScript and force them to use type assertions and other workarounds. [`eslint-plugin-react`](https://github.com/jsx-eslint/eslint-plugin-react) has a similar issue, so this plugin should avoid making the same mistake.

ESLint wants flat configs to be included in the `configs` field, alongside legacy configs.

```ts
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint/index.d.ts#L1416
interface Plugin extends ObjectMetaProperties {
  configs?: Record<string, ConfigData | Linter.FlatConfig | Linter.FlatConfig[]> | undefined;
  ...
}
```



I've seen other plugins (example [@stylistic/eslint-plugin](https://github.com/eslint-stylistic/eslint-stylistic)) solve this by creating new `"*-flat"` configs (example: [`recommended-flat`](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/packages/eslint-plugin/configs/index.ts#L33)). The legacy configs can still be used without disruption.

Eventually, when ESLint v10 removes support for legacy configs, these flat config names can be turned into aliases and deprecated.

ljharb commented:
@JstnMcBrd eslint really doesn't get to decide the export layout for other plugins, so if that's an issue with the types then that should be fixed there.

JstnMcBrd commented:
@ljharb Hey Jordan. I understand as maintainer of `eslint-plugin-react`, you prefer your export layout, but I must disagree here. ESLint does determine the expected layout for ESLint plugins. It's possible for `configs` to stray outside that layout without runtime issues, but we can't expect the types to change for every possible custom layout developers come up with!

Following the layout ESLint asks for doesn't cost much, and it reduces inconvenience for TS users (including me 😅). Sometimes there are extenuating circumstances that outweigh this, but I don't think that's the case here.

ljharb commented:
@JstnMcBrd the types shouldn't be so prescriptive, especially since the runtime allows it just fine. If eslint wants to enforce it, then the actual code that matters - the runtime javascript - needs to enforce it, and the types need to match what the code does and no more.

michaelfaith commented:
> I've seen other plugins (example [`@stylistic/eslint-plugin`](https://github.com/eslint-stylistic/eslint-stylistic)) solve this by creating new `"*-flat"` configs (example: [`recommended-flat`](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/packages/eslint-plugin/configs/index.ts#L33)). The legacy configs can still be used without disruption.

I haven't seen that convention.  But I have seen 'flat/*' a lot.  e.g. `configs['flat/recommended']`.  Honestly, I don't have a strong opinion either way.  There's a lot of inconsistency across the community for how new flat configs have been exported.  I think the inconsistency across plugins is a bigger rough edge than any one of these options on their own.  

> Eventually, when ESLint v10 removes support for legacy configs, these flat config names can be turned into aliases and deprecated.

The same holds true for any of these conventions.  

I would be comfortable with either `flatConfigs.recommended` (what it currently is and what feels a bit more ergonomic to me) or `configs['flat/recommended']`, since those seem to be the two leading conventions I've seen emerge.  Whatever direction @eps1lon wants to go.

JstnMcBrd commented:
> the types shouldn't be so prescriptive, especially since the runtime allows it just fine.
> the types need to match what the code does and no more

I agree - types should always match what the code does. But there's nuance here. I think there's a good reason the types are this way. It's not super relevant to this PR, but the reasoning is below for anyone interested.

<details>

<summary>How I understand it</summary>
<br/>

Yes, when you pass a `plugin` to ESLint at runtime, it does not care what `plugin.configs` is. If the types represented that, they would just be `configs: any`.

BUT, when a user pulls a config from your plugin and gives it to ESLint, it definitely cares what that is (see the types for [`LegacyConfig`](https://github.com/eslint/eslint/blob/main/lib/types/index.d.ts#L1096) and [`Config`](https://github.com/eslint/eslint/blob/main/lib/types/index.d.ts#L1248)).

ESLint needs to define `plugin.configs` to help devs know if they make an incorrect config. Not because the _plugin_ will blow up at runtime, but because the _config_ will, if a user tries to use it. If the type was `configs: any`, there would be no warning. The `LegacyConfig` and `Config` types would not get enforced.

Okay, so the type is useful, but why so prescriptive? Well, there are unlimited ways developers can export their configs on the plugin, which is impossible to type well. Anything is prescriptive compared to complete permissiveness.

A circular type would be necessary to represent the nesting on `eslint-plugin-react` - not impossible, but not ideal. But to represent the `flatConfigs` object in this PR, ESLint would need to assert arbitrary fields on the plugin are config objects. And that would conflict with plugins that have custom fields with other types. And I'm sure there are more ways.

So yes, the `plugin.configs` type is prescriptive, but that's necessary to be useful and define what the runtime allows when those configs are used.

It does introduce a convention, but it's fairly easy to follow it from the start (which I hope we do here).

</details>

ljharb commented:
Given that eslint-plugin-import and eslint-plugin-jsx-a11y and eslint-config-airbnb are going to match eslint-plugin-react, eslint can choose to update their types or not. ¯\\\_(ツ)\_/¯

michaelfaith commented:
I'm not entirely sure I follow your logic @JstnMcBrd.  Wouldn't the types for this package come from a `@types/eslint-plugin-react-hooks` package, in which case, the new export could just be added, in the same way that it was for `@types/eslint-plugin-jsx-ally` https://www.npmjs.com/package/@types/eslint-plugin-jsx-a11y?

![image](https://github.com/user-attachments/assets/05746ddc-3789-49da-8ba4-49ffe371cd2f)

And I would actually argue that this gives you even richer type information, and a cleaner deprecation / removal path.

karlhorky commented:
What @JstnMcBrd said resonates with me:

1. ESLint (via community-managed DefinitelyTyped `@types/eslint`) consumes configuration from the user, including plugins
2. ESLint expects this config, including plugins, to be a certain shape and match a contract (with TS types)

Both of these are pretty common in TS configuration patterns across the ecosystem.

But maybe it would be good to CC some of the folks involved in `@types/eslint` (DefinitelyTyped), ESLint and `typescript-eslint` to get some other opinions:

- ESLint
  - @nzakas
- `@types/eslint`
  - @antfu, @jakebailey, @JounQin, @linlinyang, @matwilko
- `typescript-eslint`
  - @bradzacher, @JoshuaKGoldberg

JounQin commented:
`@types/eslint` will be deprecated and unneeded soon.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70735

https://github.com/eslint/eslint/issues/18928

controversial commented:
@nzakas i believe the question is whether (from ESLint’s perspective) a plugin *must* export all configs under `plugin.configs`, or whether it’s acceptable to export both a `plugin.configs` and `plugin.flatConfigs` objects. @JstnMcBrd believes the type definitions from `@types/eslint` imply a structural constraint on the Plugin interface that would be violated by using `plugin.flatConfigs`.

basically: does ESLint have a stance on `plugin.configs['flat/recommended']` vs `plugin.flatConfigs.recommended`?

ljharb commented:
(or, `plugin.flat.configs.recommended`)

ljharb commented:
@nzakas do you mean, directly on the config object, no nested objects?

michaelfaith commented:
> We also recommend that you _not_ include the word "flat" in the config name. We recommend that flat configs just include the name (for example, `recommended`) and any eslintrc-style configs [append `-legacy` to the name](https://eslint.org/docs/latest/extend/plugin-migration-flat-config#backwards-compatibility), such as `recommended-legacy`. 

That would be a breaking change though. And this plugin just cut a major release. So, to introduce another breaking change on the heals of that, just to add "-legacy" to the existing config name, doesn't seem prudent...

nzakas commented:
@ljharb yes, that's correct. No nested objects.

@michaelfaith this doesn't have to be a breaking change. You can always publish the same config with two different names. In any event, that decision is up to you. I'm just telling you that this is the recommendation we're making to plugins for consistency.

ljharb commented:
@nzakas is there any way for eslint to enforce that?

michaelfaith commented:
Updated to have everything under `configs`.

@eps1lon, to summarize all of the discussion:

- The TypeScript types built into `eslint` tend to favor all configs (both flat and non-flat) to be attached to the same `configs` object, which was different from what I originally had done (using a separate `flatConfigs` export).  This could create some rough edges for people with TypeScript-based configs (though it wouldn't really be an issue if this package had its own types).
- @nzakas clarified that the official recommendation from the eslint team is to put new flat configs on the same `configs` object as the original ones.  
- He went further to say that they're recommending the new flat configs _not_ have any qualifiers like `-flat` as part of the name.  To rename existing configs with a `-legacy` suffix, and have the new flat configs take the old name (e.g. just `recommended`)
- That last point is the only one I didn't observe as part of my revision, since it would represent a breaking change, and with it being on the heals of the 5.0 release, I didn't think that you'd want to do another in such quick succession.  

I introduced two new configs to the `configs` export:
- `'recommended-legacy'` - the original `'recommended'` config
- `'recommended-latest'` - the new "flat" config.  I was going back and forth between `'recommended-flat'` and this, but landed on this, since 'latest' feels like a better parallel to 'legacy'.  I left 'recommended' to be an alias for 'recommended-legacy'  so that it isn't introducing a breaking change.

So, in this scenario `configs` will have 'recommended', 'recommended-latest', and 'recommended-legacy'.

If you'd rather pull the band-aid off and have the new config completely supplant 'recommended' so that `configs` only has 'recommended' and 'recommended-legacy' to match their team's full recommendation, I'm happy to make that change too, as well as write a migration guide.  Just wanted to see where you stood on that, since it'd require another major release.

smff commented:
We’re looking forward to it :) When are you going to add this?

Aluisio commented:
Any news on the approval of this?

michaelfaith commented:
Fixed prettier issue

eps1lon commented:
> If you'd rather pull the band-aid off and have the new config completely supplant 'recommended' so that configs only has 'recommended' and 'recommended-legacy' to match their team's full recommendation, I'm happy to make that change too, as well as write a migration guide.

Would that be a breaking change for people using a compat layer to get flat configs to work? Or is flat config a completely new feature `eslint-plugin-react-hooks` would support?

michaelfaith commented:
> > If you'd rather pull the band-aid off and have the new config completely supplant 'recommended' so that configs only has 'recommended' and 'recommended-legacy' to match their team's full recommendation, I'm happy to make that change too, as well as write a migration guide.
> 
> Would that be a breaking change for people using a compat layer to get flat configs to work? Or is flat config a completely new feature `eslint-plugin-react-hooks` would support? We usually prefer the shorter name to point to the recommend API i.e. `recommended` instead of `recommended-latest` but if that means releasing a breaking change, I'd rather not do it.

If we renamed the existing `recommended` config to `recommended-legacy`, it would more impact the people using the the plugin in an older `eslintrc` config.  They would have to change the name of the config they're extending from `recommended` to `recommended-legacy`

e.g.
```diff
{
  "extends": [
    // ...
-  "plugin:react-hooks/recommended"
+  "plugin:react-hooks/recommended-legacy"
  ]
}
```

Similarly, for the people using the compat layer where they're using `compat.extends('plugin:react-hooks/recommended')`, they'd change it to `compat.extends('plugin:react-hooks/recommended-legacy')` until they adopt the new flat config.  

So, it'd be breaking for that reason, but they'd still be able to work in both scenarios, after changing the name.  The PR in its current form wouldn't be breaking, but it does have the downside of diverging from the eslint team's recommended convention and a slightly less comfortable api surface.  We'd need to make that name change at some point, it's just a matter of whether to rip the band-aid off now, or wait.

samuelhulla commented:
@eps1lon @michaelfaith I think a nice compromise would be to allign the namespace with what's `eslint-plugin-react` is using since you pretty much never use this plugin without the other.

> I'll be omitting eslint/tseslint imports for sake of conciseness of the sample

```ts
import react from 'eslint-plugin-react'

export default tseslint.config(
  eslint.configs.recommended,
  react.configs.flat.recommended
)
```

Aligning the behaviour the new flat config would be under `flat` property i.e.

```ts
import react from 'eslint-plugin-react'
import hooks from 'eslint-plugin-react-hooks'

export default tseslint.config(
  eslint.configs.recommended,
  react.configs.flat.recommended,
  hooks.configs.flat.recommended
)
```

This way

- [X] it aligns with the `eslint-plugin-react` standards
- [X] it doesnt break configs for legacy users
- [X] is a clear distinction it's a flat config from the namespace

michaelfaith commented:
Thoughts on which direction you want to go with this?

rakleed commented:
@eps1lon could you please give feedback about https://github.com/facebook/react/pull/30774#issuecomment-2450444341?

Aluisio commented:
Hey guys, not to be boring and repetitive, but do we have any updates on this?

michaelfaith commented on packages/eslint-plugin-react-hooks/src/index.js:
```diff
@@ -10,17 +10,54 @@
+
+  /**
+   * 'recommended' is currently aliased to the legacy / rc recommended config to maintain backwards compatibility.
+   * In v6, this will switch to alias the flat recommended config.
```
Thanks @poteto for the feedback!  How does this sound for an updated comment?  I didn't annotate it as deprecated, since `recommended` as a symbol will still exist in the next major, it'll just be aliased to something else.  Do you think this captures that?  

```js
/**
   * 'recommended' is currently aliased to the legacy / rc recommended config) to maintain backwards compatibility.
   * This is deprecated and in v6, it will switch to alias the flat recommended config.
   */
```

poteto replied:
Yeah this looks good!

michaelfaith commented:
Rebased against `main`

poteto commented:
Thanks everyone for the patience and also to @michaelfaith for seeing this through!

karlhorky commented:
Thanks @poteto, @michaelfaith and everyone else who collaborated here!

Since I see on [the npm page for `eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) that the latest version is 5.1.0, I'm guessing the next version with this change will be either a minor or major?

- next minor: [`eslint-plugin-react-hooks@5.2.0`](https://github.com/facebook/react/releases/tag/eslint-plugin-react-hooks%405.2.0)
- next major: [`eslint-plugin-react-hooks@6.0.0`](https://github.com/facebook/react/releases/tag/eslint-plugin-react-hooks%406.0.0)

poteto commented:
@karlhorky I'm working with @michaelfaith on merging the compiler's eslint plugin into this one (new rule), so I'm thinking we'll do a minor release after that happens. It's a bit complicated to publish the eslint plugin with our current build scripts so I'd rather do it once

briavicenti commented:
Hey @poteto, any timeline on that release? I have a pending update to eslint v9 for all of my company's repos, but I don't feel great about shipping that knowing that we would (at least for now) lose the react-hooks rules. Would really appreciate this PR being released!

controversial commented:
@briavicenti the plugin supports eslint 9 since v5.0.0; it just doesn’t export a flat config. You can still add the rules from this package to your config, and they work great:

```ts
import pluginReactHooks from 'eslint-plugin-react-hooks';

export default [
  {
    plugins: {
      'react-hooks': pluginReactHooks,
    },
    rules: {
      'react-hooks/rules-of-hooks': 'error',
      'react-hooks/exhaustive-deps': 'warn',
    },
  },
];
```

guillaumebrunerie commented:
> @briavicenti the plugin supports eslint 9 since v5.0.0; it just doesn’t export a flat config. You can still add the rules from this package to your flat config, and they work great:

Except that the `react-hooks/rules-of-hooks` rule is pretty broken in v5.1.0 if you use loops (#31687).

controversial commented:
@guillaumebrunerie that bug is unrelated to flat config support, and off-topic for this PR

Aluisio commented:
Hello. Any expectations for the release date of the next release with flat config support?

poteto commented:
@Aluisio nothing concrete yet, but I'm hoping we can cut a next release after #32240 lands. Hopefully by next week.

poteto commented:
I just published https://www.npmjs.com/package/eslint-plugin-react-hooks/v/5.2.0-rc.1 which contains this fix and the conversion of the plugin to TS. 

Huge thanks to @michaelfaith for working on these improvements to the plugin!
